### PR TITLE
implement the @interrupt_task (was @peek) decorator

### PR DESCRIPTION
**Pull request by pwnage101**
This decorator turns a normal locust task into one which greedily
interrupts (to interrupt from a task means to stop scheduling tasks from
the current TaskSet instance and change context to one TaskSet level up)

---

@robrap this is more or less complete, sans documentation.  it's late on friday, so I figured i'd just post it up and see what you think so far.

### CODE REVIEW

File: helpers/peek.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+LOGGER = logging.getLogger(__name__)
+
+def is_root(task_set):
+    """
+    Return True if task_set is a root TaskSet instance, which means that it has
+    been called directly from the Locust class and so it has no parent.
+    """
+    task_set.parent == task_set.locust
+
+def peek(func):
```
Comment (robrap) [inline helpers/peek.py]: I guess I would have named this `interrupt`.  I don't understand the choice of the word `peek`?  Maybe you can explain here, and ultimately use it in the description if you decide to keep it.

File: helpers/peek.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+        @task(21)
+        @peek
+        def some_task(self): ...
+
+    """
+    def _peek(self, *args, **kwargs):
+        func(self, *args, **kwargs)
+        # Before interrupting, make sure that there is a parent task set to
+        # escape to.  Otherwise locust will crash.
+        if not is_root(self):
```
Comment (robrap) [inline helpers/peek.py]: Cool.

File: helpers/peek.py [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+            self.interrupt()
+    return _peek
```
Comment (robrap) [inline helpers/peek.py]: Nit: Maybe a line break between the def and this line?

File: loadtests/lms/courseware_views.py [hunk 4]
```diff
@@ -23,6 +24,7 @@ class CoursewareViewsTasks(LmsTasks):
     """
 
     @task(50)
+    @peek
```
Comment (robrap) [inline loadtests/lms/courseware_views.py]: I don't know if this suggestion goes with or against best practices in Python, but if we think we will almost always want to use this the way we are coding, would it make sense to combine in a single `peek_task` or `interrupt_task`?  This is a question at this point, not a request.  I can also see an argument for keeping the decorators independent.  Thoughts?

**Comment by robrap** [comment]
I'm glad this seems to be working out.  I would definitely send something out about this work when you have completed it.

**Comment by robrap** [comment]
Sorry, I pushed the wrong comment button by mistake.  I didn't mean to close this.

File: helpers/peek.py (hunk 1 above)
Comment (pwnage101) [inline helpers/peek.py]: Whoops, I didn't explain.  The idea was that a task which "peeks" does not cause the user to linger, it just peeks.  I was just trying to use a name which intuitively describes its behavior, but wouldn't be conflated with the actual interrupt() function.

File: loadtests/lms/courseware_views.py (hunk 4 above)
Comment (pwnage101) [inline loadtests/lms/courseware_views.py]: The primary reason for having task and peek separate decorators is that it allows the simultaneous use of peek and `self.tasks = {...}` to define the ratios centrally. e.g.

```python
@peek
task1():
    pass

@peek
task2():
    pass

self.tasks = { task1: 24, task2: 43 }
```

...or whatever the correct syntax would be.

File: helpers/peek.py (hunk 1 above)
Comment (robrap) [inline helpers/peek.py]: The term `peek` just isn't making sense to me.  Unfortunately, without knowing that the locust user gets stuck in a taskset until it sees an interrupt, no name will really make sense.  How about something like `exit_taskset`?  The comment could describe the normal behavior of locust and how this works.  Note that someone could also use it with a single task in a task set, the way locust documents `interrupt`, and the name of the decorator should still apply.

File: loadtests/lms/courseware_views.py (hunk 4 above)
Comment (robrap) [inline loadtests/lms/courseware_views.py]: Got it.

File: helpers/peek.py (hunk 1 above)
Comment (pwnage101) [inline helpers/peek.py]: hm, what about `interrupt_after` or `exit_after` ?

Comment (robrap) [inline helpers/peek.py]: Either of those are fine.  A comment that explains would be great.

**Comment by pwnage101** [comment]
FWIW, I did look into implementing a _class_ decorator to automatically make every task in a TaskSet call interrupt,  but long story short the TaskSet class is itself derived from the TaskSetMeta metaclass which does some pre-processing on the tasks making it difficult to modify those tasks anymore.

**Comment by pwnage101** [comment]
@robrap @tobz 

I left the removals of `return`s because they shouldn't be there anyway, but it is not related to this PR.

Users may set `LOCUST_TASK_SET` however they wish, and there are no consequences related to improper interrupt() calls (i tested with `LOCUST_TASK_SET` set to `LmsTest`, `ModuleRenderTasks`, etc.)

**Comment by robrap** [comment]
👍 Glad you could get it to work for TaskSets run as root or as children.  Thanks.

**Comment by tobz** [comment]
So, normally, as a child worker is assigned a task based on the task set/task ratios, they will permanently run that task?  And this gives them a chance to, once running the task once, go back to the "parent" task to potentially run another task?

**Comment by pwnage101** [comment]
> So, normally, as a child worker is assigned a **taskset** based on the task set/task ratios, they will permanently run that **taskset**? And this gives them a chance to, once running **one task in the taskset**, go back to the "parent" **taskset** to potentially run another **taskset**?

@tobz FTFY

**Comment by tobz** [comment]
Sure, ok, but what's the answer? :)

**Comment by pwnage101** [comment]
Sorry, more complete answer is: no, what you wrote was incorrect.  TaskSets are sticky, not individual tasks.  The structure of the LMS tests currently look like:

* root taskset (e.g. LmsTest)
  * taskset A (e.g. ForumsTasks)
    * task 1 (e.g. search_topic)
    * task 2
  * taskset B
    * task 1
    * task 2

Despite the fact that the interrupt_task decorator decorates tasks, it's the stickiness of the taskset that it is actually concerned with.  I wasn't able to put together a clean/maintainable implementation of a class decorator, but that would be nicer for sure.

**Comment by tobz** [comment]
So workers will get pinned to a taskset, like taskset A, but will migrate between tasks within the taskset they're pinned to?  Is that correct? (for how this PR works)

**Comment by pwnage101** [comment]
That behavior which you describe is how locust normally works.  The decorator in this PR allows workers to break out of the taskset.

**Comment by tobz** [comment]
Ah, I see.  So this will let us have, say, LmsTest, which has multiple things underneath it -- forums, courseware, etc -- and the global number of workers will just be able to freely roam from forums to courseware to whatever.

I just want to make sure I understand what it's being used for before I say "this seems like a good solution." :)

**Comment by pwnage101** [comment]
> Ah, I see. So this will let us have, say, LmsTest, which has multiple things underneath it -- forums, courseware, etc -- and the global number of workers will just be able to freely roam from forums to courseware to whatever.

Exactly.  So when I run LmsTest two times, the transaction distribution will not change as much, especially when there is about 50 workers/host or less.  I believe this would have helped me collect cleaner data while testing small instances cleaner, or at least would have allowed me to run fewer/shorter tests.

> I just want to make sure I understand what it's being used for before I say "this seems like a good solution." :)

No problem, you're doing the right thing :)

**Pull request merged**